### ISSUE TITLE
Three aspects on dealing with the divergence problem in single precision calculation

### ISSUE THREAD

**Issue by Simulator-CEM** [mfschubert/fmmax#93]
**Here I newly added three articles (Ref. [4-6]) about the effect of data precision in grating theory for your reference, hope it could help.**

![Image](https://github.com/user-attachments/assets/675eb302-9a54-4986-8825-817944899dff)

###############################################################################################

I have tested the newest fmmax (1.5.1), but the divergence problem in single precision calculation remains when I set jax.config.update('jax_enable_x64', False) .

 ![Image](https://github.com/user-attachments/assets/af3bbac5-e6ba-49e3-a9c3-b8d9270f79f1)

To deal with this problem, I suppose there are three aspects which may help (this is cut from "Asymmetric results for symmetric structure? #86"):

(1) FMM is similar to the coordinate transformation method in some sense. The holographic grating is equivalent to the smooth grating where the perturbative preconditioning technique (PPT) in Ref. [5] might work, and the cylinder or checkboard grating is equivalent to the trapezoid grating where PPT fails but a coordinate parameterization work still work, which is just the adaptive spatial resolution method (ASR), the detailed theory can be found in Ref. [1], and the code is implemented in https://sourceforge.net/projects/rcwa-2d/.

(2) For high-contrast grating, a aperdization filter or ramp vignetting can be used, which is equivalent to the corner-rounded for triangle grating. The detailed theory can be found in Ref. [2], and this treatment has been adopted in TORCWA (edge_sharpness for the sigmoid function, https://github.com/kch3782/torcwa/blob/main/torcwa/geometry.py,
but unfortunately, they don't use the correct inverse rule or NVM, although it might not necessary as epsilon and field is continuous after aperdization, and of course the PPT migh help here).

[3] The matched coordinate method might help [3], and the code is implemented in https://github.com/lydia1895/PMM/blob/master/ellipse_metric.m.

In fact, all the three approaches are usually used together, which might be a little complex.

```
[1] Essig S. Advanced numerical methods in diffractive optics and applications to periodic photonic nanostructures[D]. Karlsruher Inst. für Technologie, Diss., 2011, 2011.
[2] Mei Y, Liu H, Zhong Y. Treatment of nonconvergence of Fourier modal method arising from irregular field singularities at lossless metal-dielectric right-angle edges[J]. Journal of the Optical Society of America A, 2014, 31(4): 900-906.
[3] Weiss T, Granet G, Gippius N A, et al. Matched coordinates and adaptive spatial resolution in the Fourier modal method[J]. Optics express, 2009, 17(10): 8051-8061.
[4] Tishchenko A V. Numerical demonstration of the validity of the Rayleigh hypothesis[J]. Optics express, 2009, 17(19): 17102-17117.
[5] Xu X, Li L. Numerical stability of the C method and a perturbative preconditioning technique to improve convergence[J]. Journal of the Optical Society of America A, 2017, 34(6): 881-891.
[6] Xu X, Li L. Numerical instability of the C method when applied to coated gratings and methods to avoid it[J]. Journal of the Optical Society of America A, 2020, 37(4): 511-520.
```
